[Content Collections] Add `slug` frontmatter field (#5941)
withastro/astro: .changeset/large-steaks-film.md

---
'astro': major
---

> Replace the content collection `slug()` config with a new `slug` frontmatter field.

Reviewer: Can you specify that this is relevant to people using the previous experimental version in 1.x? New users trying this for the first time don't need to know about this being a breaking change.

Author: Done! Clarified this applies to 2.0 beta users as well.

This introduces a reserved `slug` property you can add to any Markdown or MDX collection entry. When present, this will override the generated slug for that entry.

```diff
# src/content/blog/post-1.md
---
title: Post 1
+ slug: post-1-custom-slug
---
```

Astro will respect this slug in the generated `slug` type and when using the `getEntryBySlug()` utility:

```astro
---
import { getEntryBySlug } from 'astro:content';

// Retrieve `src/content/blog/post-1.md` by slug with type safety
const post = await getEntryBySlug('blog', 'post-1-custom-slug');
---
```

#### Migration

> If you relied on the `slug()` config option, we suggest moving all custom slugs to `slug` frontmatter properties in each collection entry.

Reviewer 2: ```suggestion
If you relied on the `slug()` config option, you will need to move all custom slugs to `slug` frontmatter properties in each collection entry.
```
Not a suggestion? Required?
